### ISSUE TITLE
multi-echo data fails to supply TE

### ISSUE THREAD

**Issue by pfuchs** [QSMxT/QSMxT#89]
QSMxT fails on multi-echo data with the following errors for romeo_unwrapping:

```
Cmdline:
	/opt/QSMxT/scripts/romeo_unwrapping.jl --no-rescale --mag /neurodesktop-storage/02_qsm_output/workflow_qsm/sub-221207100856std131221107524367073/ses-1/run-04/nibabel_as-canonical/mapflow/_nibabel_as-canonical1/sub-221207100856std131221107524367073_ses-1_run-04_echo-02_part-mag_MEGRE_canonical.nii --output sub-221207100856std131221107524367073_ses-1_run-04_echo-02_part-phase_MEGRE_scaled_canonical_romeo-unwrapped.nii.gz --phase /neurodesktop-storage/02_qsm_output/workflow_qsm/sub-221207100856std131221107524367073/ses-1/run-04/nibabel_as-canonical/mapflow/_nibabel_as-canonical1/sub-221207100856std131221107524367073_ses-1_run-04_echo-02_part-phase_MEGRE_scaled_canonical.nii
Stdout:

Stderr:
	ERROR: LoadError: multi-echo data is used, but no echo times are given. Please specify the echo times using the -t option.
	Stacktrace:
	 [1] error(s::String)
	   @ Base ./error.jl:33
	 [2] getTEs(settings::Dict{String, Any}, neco::Int64, echoes::UnitRange{Int64})
	   @ RomeoApp /opt/julia_depot/packages/RomeoApp/X850J/src/argparse.jl:176
	 [3] load_data_and_resolve_args!(settings::Dict{String, Any})
	   @ RomeoApp /opt/julia_depot/packages/RomeoApp/X850J/src/caller.jl:110
	 [4] unwrapping_main(args::Vector{String})
	   @ RomeoApp /opt/julia_depot/packages/RomeoApp/X850J/src/caller.jl:5
	 [5] top-level scope
	   @ /opt/QSMxT/scripts/romeo_unwrapping.jl:3
	in expression starting at /opt/QSMxT/scripts/romeo_unwrapping.jl:3
Traceback:
	RuntimeError: subprocess exited with code 1.
```

and romeo_voxelquality

```
Cmdline:
	romeo_voxelquality.jl --mag /neurodesktop-storage/02_qsm_output/workflow_qsm/sub-221207100856std131221107524367073/ses-1/run-05/nibabel_as-canonical/mapflow/_nibabel_as-canonical1/sub-221207100856std131221107524367073_ses-1_run-05_echo-02_part-mag_MEGRE_canonical.nii --output sub-221207100856std131221107524367073_ses-1_run-05_echo-02_part-phase_MEGRE_scaled_canonical_romeo_voxelquality.nii --phase /neurodesktop-storage/02_qsm_output/workflow_qsm/sub-221207100856std131221107524367073/ses-1/run-05/nibabel_as-canonical/mapflow/_nibabel_as-canonical1/sub-221207100856std131221107524367073_ses-1_run-05_echo-02_part-phase_MEGRE_scaled_canonical.nii --type grad+second+mag
Stdout:

Stderr:
	ERROR: LoadError: UndefKeywordError: keyword argument TEs not assigned
	Stacktrace:
	 [1] voxelquality(phase::Array{Float32, 4}; keyargs::Base.Iterators.Pairs{Symbol, AbstractArray, Tuple{Symbol, Symbol}, NamedTuple{(:weights, :mag), Tuple{BitVector, Array{Float32, 4}}}})
	   @ ROMEO /opt/julia_depot/packages/ROMEO/3D8Ja/src/voxelquality.jl:18
	 [2] top-level scope
	   @ /opt/QSMxT/scripts/romeo_voxelquality.jl:47
	in expression starting at /opt/QSMxT/scripts/romeo_voxelquality.jl:47
Traceback:
	RuntimeError: subprocess exited with code 1.
```

The TE information should be available since the nifti conversion converts to separate files per echo, which each include the TE of the respective echo, so perhaps there is something going wrong in extracting this information from the various files.

**Comment by astewartau** [comment]
Hi @pfuchs, and thanks for posting this issue! 

A new version of QSMxT was released yesterday (v1.2.0), and is the largest update we've made in a while. Can I confirm that this was the version you were using? Was it running through Neurodesk, Docker, or singularity? You can try running `qsmxt_version.py` to see what version you are using. 

We have workflow tests on GitHub Actions that confirm these steps should be functioning, and since your issue, I have separately run the following test, which you could also try to see if the problem is related to the environment or the input data (including how QSMxT may have prepared your input data). From inside the container, run:

```
# install osfclient for download data from osf
pip install osfclient

# download and extract some test data
mkdir test-data
osf -p 9jc42 clone ./test-data
tar xf ./test-data/osfstorage/bids-osf.tar -C ./test-data

# view the directory tree of the test data
tree test-data

# run the pipeline (ROMEO unwrapping and the voxelquality should be enabled by default)
run_2_qsm.py test-data/bids-osf/ test-data/qsm
```

Thanks!

**Comment by pfuchs** [comment]
I'm running the docker through:

` docker run -it -v ~/neurodesktop-storage:/neurodesktop-storage vnmd/qsmxt_1.2.0:20221208`

which I think is version 1.2.

The test you propose runs wihtout errors, and produces a (badly masked) QSM output.

I've reran the qsm command `run_2_qsm.py` on the multi-echo GRE data separately, which proceeded without fail. It did fail on the multi-echo EPI data  I was trying this on.

Perhaps this is because that dataset has unequal echo spacing, I noticed the test set has typical even echo spacing (4,12,20,28 ms).

Here is the entire error from the EPI run
	```
 ***********************************
RuntimeError: Traceback (most recent call last):
  File "/opt/miniconda-4.7.12.1/lib/python3.8/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/opt/miniconda-4.7.12.1/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 527, in run
    result = self._run_interface(execute=True)
  File "/opt/miniconda-4.7.12.1/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 645, in _run_interface
    return self._run_command(execute)
  File "/opt/miniconda-4.7.12.1/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 771, in _run_command
    raise NodeExecutionError(msg)
nipype.pipeline.engine.nodes.NodeExecutionError: Exception raised while executing Node _romeo-voxelquality0.

Cmdline:
	romeo_voxelquality.jl --mag /neurodesktop-storage/02_qsm_output/workflow_qsm/sub-221207100856std131221107524367073/ses-1/run-01/nibabel_as-canonical/mapflow/_nibabel_as-canonical0/sub-221207100856std131221107524367073_ses-1_run-01_echo-01_part-mag_MEGRE_canonical.nii --output sub-221207100856std131221107524367073_ses-1_run-01_echo-01_part-phase_MEGRE_scaled_canonical_romeo_voxelquality.nii --phase /neurodesktop-storage/02_qsm_output/workflow_qsm/sub-221207100856std131221107524367073/ses-1/run-01/nibabel_as-canonical/mapflow/_nibabel_as-canonical0/sub-221207100856std131221107524367073_ses-1_run-01_echo-01_part-phase_MEGRE_scaled_canonical.nii --type grad+second+mag
Stdout:

Stderr:
	ERROR: LoadError: UndefKeywordError: keyword argument TEs not assigned
	Stacktrace:
	 [1] voxelquality(phase::Array{Float32, 4}; keyargs::Base.Iterators.Pairs{Symbol, AbstractArray, Tuple{Symbol, Symbol}, NamedTuple{(:weights, :mag), Tuple{BitVector, Array{Float32, 4}}}})
	   @ ROMEO /opt/julia_depot/packages/ROMEO/3D8Ja/src/voxelquality.jl:18
	 [2] top-level scope
	   @ /opt/QSMxT/scripts/romeo_voxelquality.jl:47
	in expression starting at /opt/QSMxT/scripts/romeo_voxelquality.jl:47
Traceback:
	RuntimeError: subprocess exited with code 1.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/QSMxT/run_2_qsm.py", line 859, in <module>
    wf.run(
  File "/opt/miniconda-4.7.12.1/lib/python3.8/site-packages/nipype/pipeline/engine/workflows.py", line 638, in run
    runner.run(execgraph, updatehash=updatehash, config=self.config)
  File "/opt/miniconda-4.7.12.1/lib/python3.8/site-packages/nipype/pipeline/plugins/base.py", line 212, in run
    raise error from cause
RuntimeError: 8 raised. Re-raising first.
```

**Comment by astewartau** [comment]
Thanks for providing that information! This error has not come up in our tests on real data, and equal echo spacing should not be a requirement for the phase quality step. Would you be able to share your data at all so that I can debug this further? Thanks so much! Please send to help.qsmxt@gmail.com. 

In the meantime, you can run using `--masking_input magnitude`, which will create magnitude-based masks rather than phase-based, which this step is a part of. If you would like a skull-stripped image, you can also use `--masking bet` to use a more traditional BET-based masking approach. Alternatively, you can adjust the thresholding operation to be more aggressive in multiple ways:

- Increase `--threshold_algorithm_factor` to a value greater than the default of 1.25, to multiply the algorithmically-determined minimum intensity used for thresholding (e.g. 1.5)
- Set the `--threshold_value` to a custom integer to manually specify a specific minimum intensity level if you know the range of your data
- Set the `--threshold_value` to a float between 0 and 1 to specify a percentage of the histogram range as the threshold value

**Comment by pfuchs** [comment]
Thanks for the masking tips, unfortunately I'm not able to share this dataset right now, need to get permission from some collaborators for that. I'll take a closer look at where the error originates from in the new year when I have some more time to debug this!

**Comment by pfuchs** [comment]
In the meantime I would like to note that this bug has been fixed as of release 3.0. In this release there is still an issue with multi-echo EPI data on my end but I will open a new issue with a description on that as it is related to this (still an issue within romeo_voxelquality.jl) but further down the line.
